src/CodeGen_OpenCL_Dev.cpp

```diff
@@ -16,6 +17,131 @@ using std::sort;
 
 static ostringstream nil;
 
+// OpenCL doesn't support vectors of bools, this mutator rewrites IR
+// to use signed integer vectors instead.
+class EliminateBoolVectors : public IRMutator {
+private:
+    using IRMutator::visit;
+
+    template <typename T>
+    void visit_comparison(const T* op) {
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        Type t = a.type();
+        if (t.width > 1) {
+            // To represent bool vectors, OpenCL uses vectors of signed
+            // integers with the same width as the types being compared.
+            t.code = Type::Int;
+            expr = T::make(a, b);
```

Comment:
I'd restructure this to put the if that tests if the lower down mutates changed anything above this so you can reuse the existing node here if they are. I.e.:

if (!a.same_as(op->a) || !b.same_as(op->b)) {
    expr = T::make(a, b);
} else {
    expr = op;
}
if (t.width > 1) {
    t.code = Type::Int;
    expr = Cast::make(t, expr);
}

You may have mentioned it in the follow up comments, but I think comparing vectors of bools is legal, if somewhat useless, in Halide and this code likely fails if the type of the arguments is Int(1).